From 581a3ae422350f1bba70be21c4d603b5810b43fd Mon Sep 17 00:00:00 2001
From: Tobias <ruck.tobias@gmail.com>
Date: Sun, 28 May 2023 23:07:42 +0200
Subject: [PATCH 188/196] [Fix] Fix payment requests certificates and
 logos_feature_minerfund.py test (#300)

* [backport] [paymentservertests] fix certificates and payment requests

Summary:
Procedure:

```
git clone git@github.com:gavinandresen/paymentrequest.git
cd paymentrequest
ca_in_a_box/create_ca.sh
ca_in_a_box/create_intermediate_cas.sh

openssl x509 -in ca_in_a_box/certs/cacert.pem -inform PEM -outform DER | openssl enc -base64

cd c++
make

mv ../ca_in_a_box .

./paymentrequest-create paytoaddress=1Nf4eZ76SFZdhrHVNLQkLAiS1oPPGaZxSv  amount=42 expires=1986627327 | ./paymentrequest-dump
./paymentrequest-create paytoaddress=1Nf4eZ76SFZdhrHVNLQkLAiS1oPPGaZxSv  amount=42 expires=1986627327 | base64

./paymentrequest-create paytoaddress=1Nf4eZ76SFZdhrHVNLQkLAiS1oPPGaZxSv  amount=42 expires=4826525372 payment_url=testmerchant8.org certificates=./ca_in_a_box/intermediate_8/certs/demomerchant.pem,./ca_in_a_box/intermediate_8/certs/cacert.pem,./ca_in_a_box/intermediate_7/certs/cacert.pem,./ca_in_a_box/intermediate_6/certs/cacert.pem,./ca_in_a_box/intermediate_5/certs/cacert.pem,./ca_in_a_box/intermediate_4/certs/cacert.pem,./ca_in_a_box/intermediate_3/certs/cacert.pem,./ca_in_a_box/intermediate_2/certs/cacert.pem,./ca_in_a_box/intermediate_1/certs/cacert.pem,./ca_in_a_box/certs/cacert.pem privatekey=./ca_in_a_box/intermediate_8/private/demomerchantkey.pem |  base64
```

The patch to fix the outdated OpenSSL code is:
```
diff --git a/c++/paymentrequest-create.cpp b/c++/paymentrequest-create.cpp
index 5fdbc8c..69d5b8d 100644
--- a/c++/paymentrequest-create.cpp
+++ b/c++/paymentrequest-create.cpp
@@ -178,16 +178,17 @@ bool decode_base58(const string& btcaddress, unsigned char& version, string& has
     unsigned char decoded[25];

     size_t nBytes = 0;
-    BIGNUM bn58, bn, bnChar;
+    BIGNUM *bn58, *bn, *bnChar;
     BN_CTX *ctx;

     ctx = BN_CTX_new();
-    BN_init(&bn58);
-    BN_init(&bn);
-    BN_init(&bnChar);

-    BN_set_word(&bn58, 58);
-    BN_set_word(&bn, 0);
+    bn58 = BN_new();
+    bn = BN_new();
+    bnChar = BN_new();
+
+    BN_set_word(bn58, 58);
+    BN_set_word(bn, 0);

     for (unsigned int i = 0; i < btcaddress.length(); i++) {
         const char *p1 = strchr(base58_chars, btcaddress[i]);
@@ -195,23 +196,23 @@ bool decode_base58(const string& btcaddress, unsigned char& version, string& has
             goto out;
         }

-        BN_set_word(&bnChar, p1 - base58_chars);
+        BN_set_word(bnChar, p1 - base58_chars);

-        assert(BN_mul(&bn, &bn, &bn58, ctx));
-        assert(BN_add(&bn, &bn, &bnChar));
+        assert(BN_mul(bn, bn, bn58, ctx));
+        assert(BN_add(bn, bn, bnChar));
     }

-    nBytes = BN_num_bytes(&bn);
+    nBytes = BN_num_bytes(bn);
     if (nBytes == 0 || nBytes > 25)
         return false;

     std::fill(decoded, decoded+25, (unsigned char)0);
-    BN_bn2bin(&bn, &decoded[25-nBytes]);
+    BN_bn2bin(bn, &decoded[25-nBytes]);

 out:
-    BN_clear_free(&bn58);
-    BN_clear_free(&bn);
-    BN_clear_free(&bnChar);
+    BN_clear_free(bn58);
+    BN_clear_free(bn);
+    BN_clear_free(bnChar);
     BN_CTX_free(ctx);

     version = decoded[0];
diff --git a/c++/paymentrequest-dump.cpp b/c++/paymentrequest-dump.cpp
index d20e400..d28d1dd 100644
--- a/c++/paymentrequest-dump.cpp
+++ b/c++/paymentrequest-dump.cpp
@@ -133,12 +133,14 @@ int main(int argc, char **argv) {
     request.SerializeToString(&data_to_verify);
     request.set_signature(signature);

-    EVP_MD_CTX ctx;
+    EVP_MD_CTX *ctx;
+    ctx = EVP_MD_CTX_new();
+
     EVP_PKEY *pubkey = X509_get_pubkey(signing_cert);
-    EVP_MD_CTX_init(&ctx);
-    if (!EVP_VerifyInit_ex(&ctx, digestAlgorithm, NULL) ||
-        !EVP_VerifyUpdate(&ctx, data_to_verify.data(), data_to_verify.size()) ||
-        !EVP_VerifyFinal(&ctx, (const unsigned char*)signature.data(), signature.size(), pubkey)) {
+    EVP_MD_CTX_init(ctx);
+    if (!EVP_VerifyInit_ex(ctx, digestAlgorithm, NULL) ||
+        !EVP_VerifyUpdate(ctx, data_to_verify.data(), data_to_verify.size()) ||
+        !EVP_VerifyFinal(ctx, (const unsigned char*)signature.data(), signature.size(), pubkey)) {

         printf("Bad signature, invalid PaymentRequest.\n");
     }

```

Notes:
 - the merchant certificate is the first in the chain, then all the intermediate certificate authorities in reverse order, and finally the root certificate authority. The private key is the one for the merchant.
 - do not run the `create_ca.sh` and `create_intermediate_cas.sh` more than once, or delete the previous results in between two runs. This is because re-running it causes some or all keys to be overwritten, but it fails to recreate the new corresponding certs.

Test Plan: `ninja check-bitcoin-qt`

Reviewers: #bitcoin_abc, Fabien

Reviewed By: #bitcoin_abc, Fabien

Differential Revision: https://reviews.bitcoinabc.org/D12887

* [Fix] Fix logos_feature_minerfund after the upgrade

This test has been broken after the December upgrade, and this fixes it.
---
 src/qt/test/paymentrequestdata.h           | 325 ++++++++++++---------
 test/functional/logos_feature_minerfund.py |  16 +-
 2 files changed, 199 insertions(+), 142 deletions(-)

diff --git a/src/qt/test/paymentrequestdata.h b/src/qt/test/paymentrequestdata.h
index 0f7886abd..a02587330 100644
--- a/src/qt/test/paymentrequestdata.h
+++ b/src/qt/test/paymentrequestdata.h
@@ -13,22 +13,30 @@
 // Convert pem to base64/der with:
 // openssl x509 -in cert.pem -inform PEM -outform DER | openssl enc -base64
 
-// Serial Number: 10302349811211485352 (0x8ef94c91b112c0a8)
+// Version: 3
+// Serial Number: 6D:8A:62:44:0F:AD:D9:CD:EF:C6:B0:BF:E1:68:1F:19:F0:79:CF:76
 // Issuer: CN=PaymentRequest Test CA
 // Subject: CN=PaymentRequest Test CA
-// Not Valid After : Dec  8 16:37:24 2022 GMT
+// Not After : Dec  11  10:52:09 2032 GMT
 //
 const char *caCert1_BASE64 = "\
-MIIB0DCCATmgAwIBAgIJAI75TJGxEsCoMA0GCSqGSIb3DQEBCwUAMCExHzAdBgNV\
-BAMTFlBheW1lbnRSZXF1ZXN0IFRlc3QgQ0EwHhcNMTIxMjEwMTYzNzI0WhcNMjIx\
-MjA4MTYzNzI0WjAhMR8wHQYDVQQDExZQYXltZW50UmVxdWVzdCBUZXN0IENBMIGf\
-MA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQCvua59nX9radoqDYyplcns5qdVDTN1\
-7tmcGixmMYOYU3UYMU55VSsJs0dWKnMm3COQDY+N63c0XSbRqarBcsLTkaNASuPX\
-FCv1VWuEKSyy5xe4zeoDU7CVSzlxtQD9wbZW/s3ISjgaXBpwn6eVmntb0JwYxxPc\
-M1u/hrMD8BDbSQIDAQABoxAwDjAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBCwUA\
-A4GBADSaRgK5xe47XxycXBhHhr0Wgl4pAsFsufqA9aB9r8KNEHJ0yUvvbD/jaJJM\
-RtQcf0AJ9olzUMY4syehxbzUJP6aeXhZEYiMvdvcv9D55clq6+WLLlNT3jBgAaVn\
-p3waRjPD4bUX3nv+ojz5s4puw7Qq5QUZlhGsMzPvwDGCmZkL\
+MIIC/zCCAeegAwIBAgIUbYpiRA+t2c3vxrC/4WgfGfB5z3YwDQYJKoZIhvcNAQEL\
+BQAwITEfMB0GA1UEAwwWUGF5bWVudFJlcXVlc3QgVGVzdCBDQTAeFw0yMjEyMTQx\
+MDUyMDlaFw0zMjEyMTExMDUyMDlaMCExHzAdBgNVBAMMFlBheW1lbnRSZXF1ZXN0\
+IFRlc3QgQ0EwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCuwzkQnJ9k\
+Pi00754JNf+oxcZZSmx8+25vyaV0fbtBO36e4tZcwDLC1nS9Z9PjDfLGJ/ndtA02\
+/of6SgK6NAafuj0KpSVA5jcaANfcwbEso2G/ArEI+b5cKG124eLr2SHIpVIvAK0g\
+UJW8ppA0mzIjkjuLTpXVGrJBbu05J/CgKj7SxkZHPbg50P/SyQ+C7Fs+X0pga8p9\
+QPBFugcF/zZvBpGKxPsBJ7EtZ5xbED9cqWr2/+Ktzac6nih8F9vl3YqZ0ZSTj382\
+AevrpO8PhY95zTvkehwndpiLOD0S1GtkPnaGpj74ezOFsmI5jV+bRegmtSflN4e2\
+VUcbKBfTaCA9AgMBAAGjLzAtMAwGA1UdEwQFMAMBAf8wHQYDVR0OBBYEFICgp+tR\
+grwH/JolJD6z3dauFZWLMA0GCSqGSIb3DQEBCwUAA4IBAQBGiLPMao4SW4IdcyFe\
+DTf0PSYrolAe4IPyLvUYlZXuKaGBv9ofhDZSAc7OAGLx5WLij2gi4c8gMi7pSfUN\
+I+9XEEEJ5jGM9JFcFjMPkcSX8Xq3eao6DfwBe7T8I7llxn12Fz6c8XTjDCqQ9Rrk\
+Yep3Hq9dz+eeh0Zf7SrYhOhqKhhEQK/AnCW7FY+ovjxNryNcgoMIhU2kegd97oSF\
+LHpZOnn4tBqI2tQetPC8SEDrYKQw2VetcEjCtl3kBA22dQYRCwbAzh9zL1gAPYXf\
+JAnG/l0NUCTV2a2vo9qk9i/7Cq2gdHcA7aLsi56BfVcu8yb7lcRLwQLOnPqlZp//\
+AsSc\
 ";
 
 // Serial Number: f0:da:97:e4:38:d7:64:16
@@ -60,21 +68,27 @@ z2k/tMuALCDXGeZBRPTsGHu1y4cj84swAeoDK5QSQcI+Ub7GKc+zkoj02sdDLiMo\
 // caCert1 certificate authority.
 //
 const char *paymentrequest1_cert1_BASE64 = "\
-Egt4NTA5K3NoYTI1NhrxAwruAzCCAeowggFToAMCAQICAQEwDQYJKoZIhvcNAQEL\
-BQAwITEfMB0GA1UEAxMWUGF5bWVudFJlcXVlc3QgVGVzdCBDQTAeFw0xMjEyMTAx\
-NjM3MjRaFw0yMjEyMDgxNjM3MjRaMEMxGTAXBgNVBAMMEHRlc3RtZXJjaGFudC5v\
-cmcxJjAkBgNVBAoMHVBheW1lbnQgUmVxdWVzdCBUZXN0IE1lcmNoYW50MIGfMA0G\
-CSqGSIb3DQEBAQUAA4GNADCBiQKBgQDHkMy8W1u6HsWlSqdWTmMKf54gICxNfxbY\
-+rcMtAftr62hCYx2d2QiSRd1pCUzmo12IiSX3WxSHwaTnT3MFD6jRx6+zM6XdGar\
-I2zpYle11ANzu4gAthN17uRQHV2O5QxVtzNaMdKeJLXT2L9tfEdyL++9ZUqoQmdA\
-YG9ix330hQIDAQABoxAwDjAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBCwUAA4GB\
-AIkyO99KC68bi9PFRyQQ7nvn5GlQEb3Ca1bRG5+AKN9N5vc8rZ9G2hejtM8wEXni\
-eGBP+chVMsbTPEHKLrwREn7IvcyCcbAStaklPC3w0B/2idQSHskb6P3X13OR2bTH\
-a2+6wuhsOZRUrVNr24rM95DKx/eCC6JN1VW+qRPU6fqzIjQSHwiw2wYSGXapFJVg\
-igPI+6XpExtNLO/i1WFV8ZmoiKwYsuHFiwUqC1VuaXRUZXN0T25lKoABS0j59iMU\
-Uc9MdIfwsO1BskIET0eJSGNZ7eXb9N62u+qf831PMpEHkmlGpk8rHy92nPcgua/U\
-Yt8oZMn3QaTZ5A6HjJbc3A73eLylp1a0SwCl+KDMEvDQhqMn1jAVu2v92AH3uB7n\
-SiWVbw0tX/68iSQEGGfh9n6ee/8Myb3ICdw=\
+Egt4NTA5K3NoYTI1Nhq2BgqzBjCCAy8wggIXoAMCAQICAQEwDQYJKoZIhvcNAQELBQAwITEfMB0G\
+A1UEAwwWUGF5bWVudFJlcXVlc3QgVGVzdCBDQTAeFw0yMjEyMTQxMDUyMDlaFw0zMjEyMTExMDUy\
+MDlaMEMxGTAXBgNVBAMMEHRlc3RtZXJjaGFudC5vcmcxJjAkBgNVBAoMHVBheW1lbnQgUmVxdWVz\
+dCBUZXN0IE1lcmNoYW50MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAoPhPki8Jgvme\
+O2STD9jkcnvvQZn6Y0oMjEC1DKBs813m74/JZwOPLq+zpW8ENRc5WY6l6miOASJhTHLriMdN+rGM\
+H5ZkA3yPvo9mW0pG/1hXjjl7IyMXvrhVPWTwk7pxOjd8KjTypgT8q9X4oiyvetFeU7MtQVTs7gJL\
+sE3NCybG4oGKjazDoDBcGr8GoAncT8uzU+LI7hq2k8PtNWlMO7E3ujbFnUqhHSCio+pbACsaCJ29\
+ZH8VVITsA2mYThUx4RldsDO81W1/3auy1eXgTTQIisf0x5mtw9fvdPV6lK2o7/gBO8PWAiha1hI3\
+svxWIXyfO1l1A4E+AgffOtNpkQIDAQABo1AwTjAMBgNVHRMEBTADAQH/MB0GA1UdDgQWBBSygPuN\
+tmXqiYAdYbGBrMzih19a2DAfBgNVHSMEGDAWgBSAoKfrUYK8B/yaJSQ+s93WrhWVizANBgkqhkiG\
+9w0BAQsFAAOCAQEAjrChYh+Am9MGjYJsGT4Wzv5cFJ39mOqr8JT+wyUY5wmdrHaUN0K/We0EhW9m\
+ge3H+BvbyjeHvUIdKc/a3Dby1PyM4iU+2z2TPYcLTqOin79R5VL+20z5yfsfI9wsaa2+oIwoUSOg\
+S8duILzyOKMR5VkEaQ+MNQjLPFnAB+bPbk+cEwPxVs9oqzQBMzTskavcQfUYqr0udjJsbCqlQ0z0\
+PGm6QPMxctI3NpsGskZPJ+zojuaRNPKA5Tn36Nux4QmuL3T4+gsh961ccN4kmhUqscjxYJ1SYibl\
+2hQOt1bvCSfITWlfDLYc0Owd15rrnJf2fhuwIQEm6IUGxUhKk+yX2CIxEiEIgNTb0g8SGXapFO2N\
+fiXoI85ijPezGjexnyq3JUqKiKwY7OHmnAYg/42mswcqACqAAgqWKaYGk5Yyii4BRnvPJOvShW3M\
+RDnZxdGI2mdqFbUDFY+suSOWUVUnJ9nglVIPHwoEIXb8GeUOcGzwrlKlW5X/pusCGNbOihjxpV8I\
+Vg2qPPuCQ7Ersr7/TfazxffaSnZWL3mgnh/Q2FaW8hJVG94f+6woaGRloIgyM6LiHtjSJWNtyT7q\
+DfeIWieLLCe3Y1DnN2lWKufEpUZbd7szl+MBIWmapFyLtBVGZ/+fQZeiH08XARU55SwcuMPZS0Sm\
+IkGip7u8kQStLOIDn15mSVcQQtMaX56CgLpa3i5/0tyv5tkX2CMWNBuSEZcNmokSe4OEvvIFFwOs\
+j89IvXiIrWM=\
 ";
 
 //
@@ -102,108 +116,159 @@ tejrSPOBNSJ3Mi/q5u2Yl4gJZY2b\
 // 10-long certificate chain, all intermediates valid
 //
 const char *paymentrequest3_cert1_BASE64 = "\
-Egt4NTA5K3NoYTI1Nhq8JAr/AzCCAfswggFkoAMCAQICAQEwDQYJKoZIhvcNAQEL\
-BQAwPzEUMBIGA1UEAwwLdGVzdGNhOC5vcmcxJzAlBgNVBAoMHlBheW1lbnQgUmVx\
-dWVzdCBJbnRlcm1lZGlhdGUgODAeFw0xMzAyMjMyMjQyMzFaFw0yMzAyMjEyMjQy\
-MzFaMDYxGjAYBgNVBAMMEXRlc3RtZXJjaGFudDgub3JnMRgwFgYDVQQKDA9UZXN0\
-IE1lcmNoYW50IDgwgZ8wDQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBAMMCHA3hiHbS\
-TKZ5K9jHRwE8NxkGp3IOx56PDB2diNkldG8XweTcRq7bBm7pdiBt4IVggtfs+6hE\
-hDYIOecyoAnVzPFTdvQ7KQdQ/fD9YLe6lk+o0edOqutPMyrxLFjSluXxEQyk7fdt\
-URloMMYfp3p1/hFCboA1rAsQ2RW38hR5AgMBAAGjEDAOMAwGA1UdEwQFMAMBAf8w\
-DQYJKoZIhvcNAQELBQADgYEAPsdFatnc2RJSpvZsw+nCiPVsllycw5ELglq9vfJz\
-nJJucRxgzmqI2iuas1ugwbXn0BEIRLK7vMF/qBzQR6M/nTxttah+KEu+okjps9vJ\
-cIyhfTyGPC5xkHaHZ7sG+UHOFhPw0/kXn0x+pbVgBZ5315axqcp1R+DTSj/whMAr\
-n0AKiAQwggIEMIIBbaADAgECAgECMA0GCSqGSIb3DQEBCwUAMD8xFDASBgNVBAMM\
-C3Rlc3RjYTcub3JnMScwJQYDVQQKDB5QYXltZW50IFJlcXVlc3QgSW50ZXJtZWRp\
-YXRlIDcwHhcNMTMwMjIzMjI0MjMxWhcNMjMwMjIxMjI0MjMxWjA/MRQwEgYDVQQD\
-DAt0ZXN0Y2E4Lm9yZzEnMCUGA1UECgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVk\
-aWF0ZSA4MIGfMA0GCSqGSIb3DQEBAQUAA4GNADCBiQKBgQDexUFfxb1sThvabp7u\
-dZz59ciThGmmAW0nP4tjrgEACgvWIInr2dZpTHbiQNF34ycsk0le1JD93D7Qb8rd\
-25OrpaO8XS2Li2zjR9cleixXjSLwV/zv8zJ8yPl/27XL++PDTKBXVpJ8/Syp+9Ty\
-plV1BqDhqtIHb/QSHEkTQXjeYQIDAQABoxAwDjAMBgNVHRMEBTADAQH/MA0GCSqG\
-SIb3DQEBCwUAA4GBACMooQVbkbIZ2DaPwHDc4ULwguG3VI2Kzj50UdExmHtzm2S4\
-MQei+n+HEPjtJAx5OY520+10nfuP+12H2DRLQmWmdvDpeQ/Cv0yavlw4ZRejRFo7\
-KS83C0wo5rd+qTvvOmAN4UTArWkzYcEUulPdiXnRamb0WQHTeVdIbHVkMormCogE\
-MIICBDCCAW2gAwIBAgIBAjANBgkqhkiG9w0BAQsFADA/MRQwEgYDVQQDDAt0ZXN0\
-Y2E2Lm9yZzEnMCUGA1UECgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVkaWF0ZSA2\
-MB4XDTEzMDIyMzIyNDIzMVoXDTIzMDIyMTIyNDIzMVowPzEUMBIGA1UEAwwLdGVz\
-dGNhNy5vcmcxJzAlBgNVBAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUg\
-NzCBnzANBgkqhkiG9w0BAQEFAAOBjQAwgYkCgYEAtjBRazrkebXAhXsbjimrMIRm\
-W/f9SwAHwXfc042keNtl0t2z6XE6UPcR2v/KrssXuCZgodeYxz6IM6lWosCM1xot\
-C3ChKKFBfVO30reuKBRUxXfKAFqxaG0YOAEzdZkkY9AGhqWloeSmgxpIfhInU0EF\
-JjCwrJ6IkijBatGoAAECAwEAAaMQMA4wDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0B\
-AQsFAAOBgQDBRTi1MolmOA0niHYX0A2lN5QWHkCfX0A7GwyoMA3dvM45m/NYd4WB\
-X+HwfnfYcI6X9jOgNo5OWmc4GGsld0HlxwMYEKISBS9PbSHPBrb3TBOlw5ztQpXZ\
-91+bOhLux52Fr03sK7v9qExmBM12M8UR2ltpzAMiUgLLMHyPfiWkvQqIBDCCAgQw\
-ggFtoAMCAQICAQIwDQYJKoZIhvcNAQELBQAwPzEUMBIGA1UEAwwLdGVzdGNhNS5v\
-cmcxJzAlBgNVBAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUgNTAeFw0x\
-MzAyMjMyMjQyMzBaFw0yMzAyMjEyMjQyMzBaMD8xFDASBgNVBAMMC3Rlc3RjYTYu\
-b3JnMScwJQYDVQQKDB5QYXltZW50IFJlcXVlc3QgSW50ZXJtZWRpYXRlIDYwgZ8w\
-DQYJKoZIhvcNAQEBBQADgY0AMIGJAoGBANJSH3xivX1t9olIdHsznI1aE9SD7t9i\
-SZJsIB0otoETHZRVv9M9LvyzBNK98ZV+kTOlST7PJgC0d9BQM9sgYApSRq5oqKDM\
-9FXbOm/yaReAbU3mkFNFw5roTlJ5ThEy0yOGT/DS0YBRaGIvRPRj2DiqDVdCZZ+w\
-4jo1IYHkZt4FAgMBAAGjEDAOMAwGA1UdEwQFMAMBAf8wDQYJKoZIhvcNAQELBQAD\
-gYEATm6+J1OmbrothO60xALKonWMBKr6hudb4amkFBqKbA9wMeM3jl+I/yKfz/Uf\
-xWuJ071IhiNv6Gxx5YwNvhUe1xMhUqHv0gpyK1Z47bD+kYS2se5sWNPNo3Y9qZDG\
-IXiGQxwHmrzaFk79Uy1xsmvsEz42w6hr25Yaw7HkIgrFveoKiAQwggIEMIIBbaAD\
-AgECAgECMA0GCSqGSIb3DQEBCwUAMD8xFDASBgNVBAMMC3Rlc3RjYTQub3JnMScw\
-JQYDVQQKDB5QYXltZW50IFJlcXVlc3QgSW50ZXJtZWRpYXRlIDQwHhcNMTMwMjIz\
-MjI0MjMwWhcNMjMwMjIxMjI0MjMwWjA/MRQwEgYDVQQDDAt0ZXN0Y2E1Lm9yZzEn\
-MCUGA1UECgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVkaWF0ZSA1MIGfMA0GCSqG\
-SIb3DQEBAQUAA4GNADCBiQKBgQC7vVUFpxHzz2Tr/xij3k58s8d/BPA0R6D5RXTV\
-vmhAzc1Zuin4zUKRFs/aCj/0yED8Wu/COfNGF4tVlRNMdl9EcFsxa8XGEL4eAZa+\
-H/rOHH+7/1EINrrVWhZlUecyhilN8jmCZmqEM3ecuD0NAViqyMrgmaiFmsLoQZpE\
-GepDUQIDAQABoxAwDjAMBgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBCwUAA4GBAEdJ\
-Ss8jWiooja3WZzHXeF95QkBJNjIlpDLGcpl4opOYLSuEl9Uxp//LaQQiXuzpj4/I\
-pkWGQmMy5HOyH1lqDyiMgXpcG8PE0jEQAoEUGZ0QEqB1mZ6BCrYvmUuf/5aSVd8Y\
-6lKMR3WzFDYU9Zy0nzuHB/3nvp6MeDRQeRMtYvz4CogEMIICBDCCAW2gAwIBAgIB\
-AjANBgkqhkiG9w0BAQsFADA/MRQwEgYDVQQDDAt0ZXN0Y2EzLm9yZzEnMCUGA1UE\
-CgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVkaWF0ZSAzMB4XDTEzMDIyMzIyNDIy\
-OVoXDTIzMDIyMTIyNDIyOVowPzEUMBIGA1UEAwwLdGVzdGNhNC5vcmcxJzAlBgNV\
-BAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUgNDCBnzANBgkqhkiG9w0B\
-AQEFAAOBjQAwgYkCgYEAxYYo3w2UXiYg6O8b4QgwN/vgreTkiW122Ep/z2TiDrhV\
-MhfOOiKdwYESPflfnXnVaQQzCGexYTQqsvqvzHSyna5hL0zPTRJxSKmTVrXRsWtp\
-dCRhjxCGipS3tlQBDi7vb+7SNRIBK4dBjjGzALNk7gMCpy+yM8f6I043jTlmGb0C\
-AwEAAaMQMA4wDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQsFAAOBgQDU+IQxt3Oh\
-KqaUYWC23+cB2gekvWqwMBnrCNrX/Dp+kjoJKUoR2Fs3qw53raHES4SIhpGT9l9l\
-rppNQgFe/JMHeYqOZMZO+6kuU0olJanBJ14tPIc7zlMTQ9OfmZ6v07IpyFbsQDtR\
-hpe80DpuvSFPfJ4fh0WrQf6kn3KDVpGDnAqIBDCCAgQwggFtoAMCAQICAQIwDQYJ\
-KoZIhvcNAQELBQAwPzEUMBIGA1UEAwwLdGVzdGNhMi5vcmcxJzAlBgNVBAoMHlBh\
-eW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUgMjAeFw0xMzAyMjMyMjQyMjlaFw0y\
-MzAyMjEyMjQyMjlaMD8xFDASBgNVBAMMC3Rlc3RjYTMub3JnMScwJQYDVQQKDB5Q\
-YXltZW50IFJlcXVlc3QgSW50ZXJtZWRpYXRlIDMwgZ8wDQYJKoZIhvcNAQEBBQAD\
-gY0AMIGJAoGBANzgVP99Qg98e6NsKEz1v5KqRB7NTBRRsYnBvb/TSWipvMQaCYuE\
-yk1xG57x++QuASKeR3QHRQJOoAhQaj9JLUhSSv9GQ5PrFLLsOFv7L1tpzXHh2dOB\
-IW92X2yFRW2s39q+Q21yvN+N8uoKdqXhzRA+dDoXh3cavaVeHX1G+IrlAgMBAAGj\
-EDAOMAwGA1UdEwQFMAMBAf8wDQYJKoZIhvcNAQELBQADgYEASTwg84cX+1UhOG9s\
-ejFV3m34QuI1hPZ+qhqVJlRYUtego8Wng1BburDSwqVAv4ch2wi3c2s4e8J7AXyL\
-tzSbSQG4RN0oZi0mR8EtTTN+Mix/hBIk79dMZg85+I29uFA6Zj2d9oAhQv2qkHhc\
-6tcaheNvkQRlCyH68k3iF1Fqf+4KiAQwggIEMIIBbaADAgECAgECMA0GCSqGSIb3\
-DQEBCwUAMD8xFDASBgNVBAMMC3Rlc3RjYTEub3JnMScwJQYDVQQKDB5QYXltZW50\
-IFJlcXVlc3QgSW50ZXJtZWRpYXRlIDEwHhcNMTMwMjIzMjI0MjI5WhcNMjMwMjIx\
-MjI0MjI5WjA/MRQwEgYDVQQDDAt0ZXN0Y2EyLm9yZzEnMCUGA1UECgweUGF5bWVu\
-dCBSZXF1ZXN0IEludGVybWVkaWF0ZSAyMIGfMA0GCSqGSIb3DQEBAQUAA4GNADCB\
-iQKBgQDaV8zhfyQuSf/f+fauMfgs3g/RnWy9yxxUkvQneQQPH3uZzCyk3A6q72ip\
-TtwNqiibG9455L9A7SaUjGtnpUz0NKT/VWUdqbfCl1PqXjEZbDobbAQ5hxLGOTyL\
-RQhLIcgeq2/BnmeCqHsC4md04nUp+nBo1HwKyygvK+9sMbCp/wIDAQABoxAwDjAM\
-BgNVHRMEBTADAQH/MA0GCSqGSIb3DQEBCwUAA4GBACvYyE+PPmWFkbjyRu9LAt8D\
-crtyYYLRClKSg6tVvutwukLG2l//kDOohYkJtgTqr6LnCIIIwYdXN+4wxugmw4cn\
-PIZmP6kovxjhhVM95okilor1zniTAo3RN7JDIfTGNgxLdGu1btt7DOFL4zTbeSJM\
-b8M1JpPftehH+x/VLyuUCuoDMIIB5jCCAU+gAwIBAgIBBTANBgkqhkiG9w0BAQsF\
-ADAhMR8wHQYDVQQDExZQYXltZW50UmVxdWVzdCBUZXN0IENBMB4XDTEzMDIyMzIy\
-NDIyOFoXDTIzMDIyMTIyNDIyOFowPzEUMBIGA1UEAwwLdGVzdGNhMS5vcmcxJzAl\
-BgNVBAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUgMTCBnzANBgkqhkiG\
-9w0BAQEFAAOBjQAwgYkCgYEAo5Vy9H3nA/OOkF5Ap89yfVNSiTay/LYCaB0eALpc\
-U690U75O9Q3w2M+2AN8wpbbHsJHZMIjEeBRoQfjlYXW1ucQTxWKyT+liu0D25mGX\
-X27CBXBd4iXTxVII/iX+u3lcjORjoHOBy7QgeIDIIS9y0vYu8eArpjh7m4thrVgI\
-RtMCAwEAAaMQMA4wDAYDVR0TBAUwAwEB/zANBgkqhkiG9w0BAQsFAAOBgQB9LKcV\
-JK9sjASNzpQlpUp7nCiw5FSjVY+XMRIKK/kavzlKjZ+InsmmyRVGjDoZi9GrqG9P\
-VHgLBxi2VtVjmokZoNPqao3OfhqORAubC+JR/JLepM7aDaxDdTHVhSUk4lgNAvi2\
-6dGY7nZMsnHlPQ2tPp/HvRRiMq1oDjlylc8VTCI2Eh8IsNsGEhl2qRSVYIoDyPul\
-6RMbTSzv4tVhVfGZqIisGLLhxYsFKg1Vbml0VGVzdFRocmVlKoABn2HTsUQtMNI4\
-yNvkfkFNka3pRvTUTydJrvyfmEeLzImfM1BWddZjnywku9RToNFZZNgow5QnljmF\
-chhR/aHOuEMTxmc12K4rNlgYtHCsxLP9zd+6u0cva3TucZ6EzS8PKEib/+r12/52\
-664NuWA9WtsK7QCFrK2K95PnVCRmWl0=\
+Egt4NTA5K3NoYTI1NhqvQQrEBjCCA0AwggIooAMCAQICAQEwDQYJKoZIhvcNAQELBQAwPzEUMBIG\
+A1UEAwwLdGVzdGNhOC5vcmcxJzAlBgNVBAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUg\
+ODAeFw0yMjEyMTQxMDU0MzJaFw0zMjEyMTExMDU0MzJaMDYxGjAYBgNVBAMMEXRlc3RtZXJjaGFu\
+dDgub3JnMRgwFgYDVQQKDA9UZXN0IE1lcmNoYW50IDgwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAw\
+ggEKAoIBAQDN7BrMYqRxRCN2yOiFA3+akEAYo6FwU7sa9T0E2CTEDgP4B2P9RZJd8CJZSiwOOdLC\
+Aq4bZ4Av/b0VSb1KdXw5WV1orbm4gm4lqstMTnilMfve7bIEtcohp4NKZI3pIJaNmdIMTNbRQFio\
+iAs9gyasoy/BdapzAODGRuWJ0AZDsmTJGXfPpgZwqXBkE2lM+Vl+gp27uMNvF4R5Qhrdx+MD8Hv1\
+bJEuV89fGHeKSfGUiDs0aoLau6QcP4209L0EF5trosi2Q67Sb/dTVvnwlnzWazfLDUH7mtqmoZp1\
+q1VC4Msj168P+Nq9KTPOL7Lwwu8S23Le0FUt3btfageGb71NAgMBAAGjUDBOMAwGA1UdEwQFMAMB\
+Af8wHQYDVR0OBBYEFCDvdEjWxylaGBsbZuXF3xrsoql4MB8GA1UdIwQYMBaAFCHVEBTn6TJlDXRf\
+0wuicIuoej16MA0GCSqGSIb3DQEBCwUAA4IBAQDMAo43e5kzonApVWwO87hC6BAhzE6RHpuGlIkm\
+cEPEHHVW8tN1ToqsxiXaFxCu8GRqFz9/uNsrrSL+USOl/s7ZPNkYkmJcwIi7g13bVcS3L3VBa9N5\
+dpgwHBsjg2oHwYqodE/wGfe/kLrrCqzhr2itPwZc5vGvnYE9+Wp9APD/JdpwmpZ6xFkYCT2Iz5ID\
+9hdrQZHCR5mYC2jJAFR1IYoPhvQI3jClyPvetIPLk7HC3J1I2E7jgFlfgyFrXLz/P8xUNhOENLha\
+YVSvGYCesCUUaimXTcaVb4MOeB3nceDUrIUbWvJi6qukEbSNXh2/ZL7pMW0uN2JrmBVeH6qIgmrZ\
+Cs0GMIIDSTCCAjGgAwIBAgIBAjANBgkqhkiG9w0BAQsFADA/MRQwEgYDVQQDDAt0ZXN0Y2E3Lm9y\
+ZzEnMCUGA1UECgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVkaWF0ZSA3MB4XDTIyMTIxNDEwNTQz\
+MloXDTMyMTIxMTEwNTQzMlowPzEUMBIGA1UEAwwLdGVzdGNhOC5vcmcxJzAlBgNVBAoMHlBheW1l\
+bnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUgODCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEB\
+AOZqb6jQy3mibfFHhaw2+csH7EAs+S3n/rbZZptKNPBMUNgVp4a8Lw/sOsMNGsliYbt1k39xroJi\
+NHot637qvnD2B8pH9uB6t4TjxE06xpgh8QhMTQ5s9y0yxJH5p7ltAftt2OZjUrG1iaQNwcjzccho\
+Q1xhQ/fFI3ywQoF6yx3AsbUgV5OmFv54WPqXneF4omPc7z43HzGdQ5TJw3VMbapcIm/IgC80E5PO\
+Ni+uVAsB/DDd6YT0LOQw9LzDXxaZB2mLwyBsNm+X0g3A0G6f2n7ozBRl9Qr2O+nXgwzf1ftM385A\
+sQfBMyBPkwrhphY8Xj5Tayi3QBP417zfUvA+pFkCAwEAAaNQME4wDAYDVR0TBAUwAwEB/zAdBgNV\
+HQ4EFgQUIdUQFOfpMmUNdF/TC6Jwi6h6PXowHwYDVR0jBBgwFoAUGz1zdRgA18OA4+TfzWLLRlkG\
+fSYwDQYJKoZIhvcNAQELBQADggEBAFQdUOTVnMog8dMjSKsbJ+65KLpO71Umspseqq3bHKkCZtXh\
+V/OhR9Q0YOOWGiOG8LPMEvRjDSVxXNrkS6e/hgHavUOXrh8MgHzCOOigGXP/Y1kubLztrFzhzknI\
+pqGEeE3zn8QyrARrZqgRjvdNGXs/R1iBQZZp5bUdLD3guJ2vgOqaIVxpROSYMM5HMInfKdOkrkPB\
+YmsxkPrq3T+Sc4tEi/wLy3OZr7Nvbc16FbtaO3A6NkozXX4oh3dYHxPl6jsaF+BDqPsOdxAJsP+e\
+AriM2/HaOrDwytz32aLPB78gfgOLIorVSlZinSgyfsKbXzIF5QWbHUnF98N3emMvw0MKzQYwggNJ\
+MIICMaADAgECAgECMA0GCSqGSIb3DQEBCwUAMD8xFDASBgNVBAMMC3Rlc3RjYTYub3JnMScwJQYD\
+VQQKDB5QYXltZW50IFJlcXVlc3QgSW50ZXJtZWRpYXRlIDYwHhcNMjIxMjE0MTA1NDMyWhcNMzIx\
+MjExMTA1NDMyWjA/MRQwEgYDVQQDDAt0ZXN0Y2E3Lm9yZzEnMCUGA1UECgweUGF5bWVudCBSZXF1\
+ZXN0IEludGVybWVkaWF0ZSA3MIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEAzBDV5IhF\
+LYqZfH4vtoCH6rhihy8ARGHHQSj5HxxiIIZMWNmG4HcOB0b43LA+LRXPgKGnrFQcj4+5vsxCTAln\
+HzgMJhET3R6g1AqaNTx7Ku/xq2ndb+TLUaRpP/32h1kMrMzesEWO/6d1QNOYgi7UPVFx+m+4S7su\
+aCCcwUILxslmEloXBG9yw9Ue+aJN+9VVJv1w+vFJMgowTHXP8KNgLVBJfTFTkgMMeGtdBudjKGLq\
+lIcD8in0o4yjRsiW+C++0JQ7gdxCTHpkqg+Zn7t4ih/qCwGK6IykS7uSr57AjF0D/5c/dQx+Zoeu\
+Ch9JWtRmgB+CQ7hlPjJBrYcR3L7qoQIDAQABo1AwTjAMBgNVHRMEBTADAQH/MB0GA1UdDgQWBBQb\
+PXN1GADXw4Dj5N/NYstGWQZ9JjAfBgNVHSMEGDAWgBQGEjMn/9OvzWjzGXb7WRHFhORhMTANBgkq\
+hkiG9w0BAQsFAAOCAQEAnmDL/bvc9QvZvzKI/smM+02frY9SIPloOL2XAj9WdRMV9sKUIur4n26Q\
+gxBimofFduzM3r/2V5EqCRDGJ6ZhYbsI6dhghSV0enKs6/lDe/aMQfXeE8mArgS8mkWBvSeXTJ2N\
+T5mFatE1JAmF+k3n1niA4M52EPUA3dX+oxmeD9zxlPWQM1s/gHeVrnUlMkURr8pPxOnMeAYyvIOJ\
+aZguEkZS58sngZEmu7lpNV4K2b9bB3cU3qHnWKp2eu4y+owKfmrPFIAWFEW0DxxQqVj9kXZ7Hgvj\
+ls09lji4nsor/+2AbLLBW0yMIJtfEVqNA61oZlg5LfmsUDS8almm/V2bGArNBjCCA0kwggIxoAMC\
+AQICAQIwDQYJKoZIhvcNAQELBQAwPzEUMBIGA1UEAwwLdGVzdGNhNS5vcmcxJzAlBgNVBAoMHlBh\
+eW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUgNTAeFw0yMjEyMTQxMDU0MzFaFw0zMjEyMTExMDU0\
+MzFaMD8xFDASBgNVBAMMC3Rlc3RjYTYub3JnMScwJQYDVQQKDB5QYXltZW50IFJlcXVlc3QgSW50\
+ZXJtZWRpYXRlIDYwggEiMA0GCSqGSIb3DQEBAQUAA4IBDwAwggEKAoIBAQCwyB5C4uZwnZLNzLE4\
+ZYgw+QZbLZDDiC2wJRdJ52BRJVwbj+xXVghr662bh+SRNk3QC6CO8M1MZdfO10BI6vt89jZpEapE\
+0ZX82RFt6LiifcT7+WqdgTzp//X9nm+do29vgaiMjb+MG18HdtbCw3hQa9UcffydkHspmG7B9DDB\
+TpWZNxdO7C6yiQvemwR89oiyclnnuzMdCa4VNzi7g0KI81bZ+G6wKNVLfGUDDVjeDJhWnS6DY/1p\
+ivhmDMg68/x4i3BS3yI3PZAnpAQuJCSwoo42ZsTnTqyw/T1g5W9qD+JBnrxI0jqzAtcPwuucA+eN\
+rqRu1PlcvXNpkOnf6J3FAgMBAAGjUDBOMAwGA1UdEwQFMAMBAf8wHQYDVR0OBBYEFAYSMyf/06/N\
+aPMZdvtZEcWE5GExMB8GA1UdIwQYMBaAFGnQiDqva6EaQ999TgYmnhm7ubeyMA0GCSqGSIb3DQEB\
+CwUAA4IBAQAVFS/V3zWbO503jkrfDv+ueuSsbDOcGYL7tQe1qXT5IkxjSaP1uqUyfpFgD0ngXfyi\
+NOkV+iH6psNprh6mpxNKKcDzB1oVyrxp7zYWOhro5g5yLwGmsp9H7OA9vw3IcB7iqbBDiSfAATwF\
+zncTmgHJUxWI4bu+9itbXPu0Z7WV7P27DsQe2qpVQ+OikDIU9mdVILSQSIh5GUVPweH2QyVXlh4o\
+3EhkzdjVFRx9hYpHSRxEV6BzSvxtBWRFZhLW4sQ66/9xjzDgnOsD0QmcCkGmP4/P+zrp7DMi4srY\
+HapRdRIzFTq+K9FHQv78KcU7SjoFszuB0Gy+iwn/3fW7WzDxCs0GMIIDSTCCAjGgAwIBAgIBAjAN\
+BgkqhkiG9w0BAQsFADA/MRQwEgYDVQQDDAt0ZXN0Y2E0Lm9yZzEnMCUGA1UECgweUGF5bWVudCBS\
+ZXF1ZXN0IEludGVybWVkaWF0ZSA0MB4XDTIyMTIxNDEwNTQzMVoXDTMyMTIxMTEwNTQzMVowPzEU\
+MBIGA1UEAwwLdGVzdGNhNS5vcmcxJzAlBgNVBAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlh\
+dGUgNTCCASIwDQYJKoZIhvcNAQEBBQADggEPADCCAQoCggEBAIyYAqLJA6yOk+cZQnyMjNEXutug\
+pYPY6tSbTlsBsrM/X/lYOjDqdmpFATrd7Kqr8jwMVu6FB8PVERnddluejIDzoMrpHqGZnqzgsX4X\
+GNYo4c4WUEZlq9R/0vwnqLp7ZNwC+X3+AA0liyLEhMOJ6Qp+D9KEnrq7TjQL9DA/dEoKtmdHFs8o\
+LDRS7uDfCk507Aa791aVSgwL9MaMmpnstp2AnFo+40ML4a9ToR9a47fmFTQse8zh88THK6wD0nhe\
+FHunX4ARrfG/qQ1yxC/Xg0RoKBudw+RKZ1BKWBEUpAMBeTIwZDTV7y4iIJtFwZErb7C2iPodUPkC\
+an0N7ZlrjcECAwEAAaNQME4wDAYDVR0TBAUwAwEB/zAdBgNVHQ4EFgQUadCIOq9roRpD331OBiae\
+Gbu5t7IwHwYDVR0jBBgwFoAUixqIAFD9Ik7qaxCIm1qhpZMHYtUwDQYJKoZIhvcNAQELBQADggEB\
+AEkU4UxV0HaZW9XHyh3zWcr9NCDOd8TKRgDdgWXsDEgWk7+tTwbU3e4HufkFkJzHOYoCAyQ2oM/V\
+JRgmZNchf98YBs3ovMLyJoGZP7fScCe6jhReUnjDWd+1ZRNKzPTLffO0/CVQCcYpP+aOQ0fpaacN\
+6TAD+79Thw4L0jgd/nyuwwN+VbaCv00ZnvkDLzirSe0Mv8ndwlPK3SrUs3oLhoyCOGHZrf4pBFZf\
+lg3vIIFcP+liByq7pc9hQwVuTsPe/FkGzG9Q1iLIA8uz+cMBrC3ALGOgCArD2nWPSh90fQ+nAKWo\
+w1uaCMZYwuPEcfmiGz1JV0O08DwD74anNIw6RyIKzQYwggNJMIICMaADAgECAgECMA0GCSqGSIb3\
+DQEBCwUAMD8xFDASBgNVBAMMC3Rlc3RjYTMub3JnMScwJQYDVQQKDB5QYXltZW50IFJlcXVlc3Qg\
+SW50ZXJtZWRpYXRlIDMwHhcNMjIxMjE0MTA1NDMxWhcNMzIxMjExMTA1NDMxWjA/MRQwEgYDVQQD\
+DAt0ZXN0Y2E0Lm9yZzEnMCUGA1UECgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVkaWF0ZSA0MIIB\
+IjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArT+UF08069BaYj94eZOhHVPOtnZziENVMp0G\
+FtFWUeXngj9fk+O+Qk9qL6cf9OlNZEdsTEWf96NJcnSeGJuNcDoYbBoGhFO6YRI+jLNgfCPcyxBY\
+5lwq8cv3G2VAPIwu1BvNhvii3UtPTj1AEutygacOK7D7drKxoaqJolPp8nBinoo1rzncIiXpDEIs\
+V+KJgkUPgXtNE/XEdt4LesqCNQECSxF9vidmokg7ht54e3Pz5hcZee4QGi6INEBGx4k052YBCeJm\
+Im81FIZpJH5JxAY+f6pd5UTavJlO5s6GQGKndmTmgrIcAFp5JP7ZkT8FjP/8N0V80fxiatFCT3wi\
+CQIDAQABo1AwTjAMBgNVHRMEBTADAQH/MB0GA1UdDgQWBBSLGogAUP0iTuprEIibWqGlkwdi1TAf\
+BgNVHSMEGDAWgBSRhZuPf+OfO/H3Q+mB8nudt2lAZTANBgkqhkiG9w0BAQsFAAOCAQEATC8BGYPB\
+TGo3ne7+zjzw74lWx42c9LhIRZqyNbFJuy+fiOcBGWQgzE4UGfsoEgSH6b1vdnz4sYGpg3+61B+3\
+tFqQX+a2+9f0HmGof/uxQL0LGx9ObUDoNkdRcaWoA4jdZaTRDAvAYQ/bhPbRAtCmlPq+ddZMQIqM\
+ILwDT0t4h20xrxsyjEdchQiwvW6hHurPbwK/R2vWBzTVNytxO/SZz87UrFw9d9FCdu8CtEYnoJ3N\
+BqkCYrkcw9Njni8yrzyknTnGyuWWpNfQYBoFM2Kcquf5KsVnNd6XrcXgrxerQ+IaE+4RlaCZdXjy\
+gRhzAyz3YnC8d9lDYH5V00ac1TeefwrNBjCCA0kwggIxoAMCAQICAQIwDQYJKoZIhvcNAQELBQAw\
+PzEUMBIGA1UEAwwLdGVzdGNhMi5vcmcxJzAlBgNVBAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1l\
+ZGlhdGUgMjAeFw0yMjEyMTQxMDU0MzFaFw0zMjEyMTExMDU0MzFaMD8xFDASBgNVBAMMC3Rlc3Rj\
+YTMub3JnMScwJQYDVQQKDB5QYXltZW50IFJlcXVlc3QgSW50ZXJtZWRpYXRlIDMwggEiMA0GCSqG\
+SIb3DQEBAQUAA4IBDwAwggEKAoIBAQCml/oi7Yv3gYtRXeczHZ5MEge0AOOrXi7xabufoz4KOxV8\
+AU7elfsY0etQ3vxXkQxW1GF8D03aV5kHbc6XDW25BZUpF9oV7B0EhcajGW/Z9VcqhWChKup7dWz9\
+yqBFKU6ZIo4PdV9iwSR58xoO4QWM72vEg+/URfPcs/eb5uuOQGewi4RFeK4r19Y8oewFTkHT8+7z\
+n3wz+zu/RoVvHSmknbGDHD+CGwhb4Tx8VlWJ9w1EPMSUxdvDb7pE2FUme52+XMhNaRiTL+K05nPu\
+iiRREwTFd5oMz59dpzquPmQyfOZiM1udce1pjOsVafb68YB6lE418NaBu36RzB39dTwLAgMBAAGj\
+UDBOMAwGA1UdEwQFMAMBAf8wHQYDVR0OBBYEFJGFm49/45878fdD6YHye523aUBlMB8GA1UdIwQY\
+MBaAFDkYXuqRU3HtVQY+n6eOKd2YldEsMA0GCSqGSIb3DQEBCwUAA4IBAQAyTzMu/xzsuRdiY8Mm\
+Z5+eqGfisfBvvLrwChuyWRw8FeCG7jn74+tVSVEQwOesvk+SLZN3Q/HPIxYqhUHFKQMB7Rr6c/7M\
+IJOB5b7YsR7HDkTWz16alOvXa2LN+j3jEf/yfPF/TZkbG3dN69WNlKXnvvPOpCIN5M26LkSGflE4\
+mWu7m7UxMIpgz/U2DkpqHMaCYDMmt2riRTOuDxlOWFHz139LtPhuStHJ4tSeciWCUhaDEVx8CcoK\
+wMvluyBsRgUVrfytV9fLvD9MmzR0bzZKOEQ4PTrlFqsVop+YwOJcbLQcRdmVMMZ5tv5s3eoUS0Bs\
+7c10mG0lxFXDjvX6y42qCs0GMIIDSTCCAjGgAwIBAgIBAjANBgkqhkiG9w0BAQsFADA/MRQwEgYD\
+VQQDDAt0ZXN0Y2ExLm9yZzEnMCUGA1UECgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVkaWF0ZSAx\
+MB4XDTIyMTIxNDEwNTQzMFoXDTMyMTIxMTEwNTQzMFowPzEUMBIGA1UEAwwLdGVzdGNhMi5vcmcx\
+JzAlBgNVBAoMHlBheW1lbnQgUmVxdWVzdCBJbnRlcm1lZGlhdGUgMjCCASIwDQYJKoZIhvcNAQEB\
+BQADggEPADCCAQoCggEBANQyBNrOC2Ep6KEENggBAnYvYbB35Mk9PpQgRi0bv4kVhOtTxWDM/Vd2\
+lzK1oQ95O8dv6WNnnkBjqlVg3KiQWOU+7EBONngShM/H+RSmERZ+LuKPl4nc0zKoVqSsj0YhT5sF\
+qelEcaZsFJ0PRahhbTnCrtpVSgs+RDPZqcW5ibS44INXYmqsM9PfYsmuLaPmOZKfPDJw3YgEsa5F\
+h2dZOqndyjrVXvYOqGLTtxKDXFyDmee50mWvTuf+VU7CCiWvSHZiK/bdKDs3XAMqdnghKPmEimOf\
+YmrD6mbA48uNTZYtQpCLkAEKRUpIxz8U7GFuogLv8h3hl8YUaGNi6ukVK7UCAwEAAaNQME4wDAYD\
+VR0TBAUwAwEB/zAdBgNVHQ4EFgQUORhe6pFTce1VBj6fp44p3ZiV0SwwHwYDVR0jBBgwFoAUm1sH\
+8rcVh/qWnExA7RoNTcpcmpowDQYJKoZIhvcNAQELBQADggEBANbfQrftMNogAQTuGmNlH/Vj0c3M\
+IXvE6T6+Q40pCR54MmfUEeol7L/UBkxgG/lrZN5GBSeHKUGENqKj0J1+blGELBBWETwKZzLJWNjE\
+gFoVJ7XtqggxA6g4+PL4r2yvdFiFTvpfVX5b6ycAneyfp42eZdOfy7WKTJbHryLxsidgiXAlZrvB\
+K5pOeARwK8TuVuRJu7lm+FP9DAJ+chfY/2oX4y3mTXuZ1X8rCdCdHxjcV9OiXFYlXEXaO92o8RH0\
+d2RqQNZZzc8b9ki1crlIAj3Om+biB6eurfPMCyeK0IPGDdif5KyLYkX9416Km6TCDHjWxeDAE+OU\
+6+xtJJZFww0KrwYwggMrMIICE6ADAgECAgECMA0GCSqGSIb3DQEBCwUAMCExHzAdBgNVBAMMFlBh\
+eW1lbnRSZXF1ZXN0IFRlc3QgQ0EwHhcNMjIxMjE0MTA1NDMwWhcNMzIxMjExMTA1NDMwWjA/MRQw\
+EgYDVQQDDAt0ZXN0Y2ExLm9yZzEnMCUGA1UECgweUGF5bWVudCBSZXF1ZXN0IEludGVybWVkaWF0\
+ZSAxMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEA4HQnuNA1qsDR1JsFq3Uab+ESxkst\
+b4Ck9TjlfRPMToZuc6C3hm6wKIM4PmKLtIDe66npoyrbxdMVhu9Mz0OTNT81n9YxB/mShCIPiw+f\
+PALK+4cDC6iVL/HtrlkzGct1gNVVMo3TAsM+3Ey7HCNPPp1DMO7nvlFhqFkJ7ei+h/n43gF/wQHY\
+Fv9rIwTw+3ZW/HxZafWofsCdfPIH9kwhvEpmWBIkVEFqElWpDE8v0jHFDCvZ4BhgdtO5O0h0cTwx\
+KjpZzgEJNIQPXYDTftZm6i9VrDQbkfiPy0BzoHOqxxtJqiyg38Pr5z5Gj8GBRoHm1sbazHRm7QkW\
+FsjiYdhZLQIDAQABo1AwTjAMBgNVHRMEBTADAQH/MB0GA1UdDgQWBBSbWwfytxWH+pacTEDtGg1N\
+ylyamjAfBgNVHSMEGDAWgBSAoKfrUYK8B/yaJSQ+s93WrhWVizANBgkqhkiG9w0BAQsFAAOCAQEA\
+MKdLTcMUklnmDDVvoJ5seGuoWbO+xOHTd4E78wUJHrfR5+4IFFzGnrsVqR7CJ5t6P/EElkp1q0po\
+EOjXieV7NKIszym13L9IMv6eiIPNKLlHWFGTbo6JEEkVRR/NeYN4i8O4+ibbIQyyLZnD0rQ8QWYf\
+Bk6s7XtvStIWwtHH0K01qwzp5WPFFunZKureTDEv8QV2baSADq4b1HC1MAeIb2M9z+0JZDHzN2ir\
++xlyd+PLIICDFz7xbrJsWjC2IACtntgiJczGL39CNE9dG8vLJ4kwVH3pg3HC6638f9hFXzfO1nSd\
+aezfFhbXfhY2UuxQhitUPCsgd0BZRsNbHhatKwqDBjCCAv8wggHnoAMCAQICFG2KYkQPrdnN78aw\
+v+FoHxnwec92MA0GCSqGSIb3DQEBCwUAMCExHzAdBgNVBAMMFlBheW1lbnRSZXF1ZXN0IFRlc3Qg\
+Q0EwHhcNMjIxMjE0MTA1MjA5WhcNMzIxMjExMTA1MjA5WjAhMR8wHQYDVQQDDBZQYXltZW50UmVx\
+dWVzdCBUZXN0IENBMIIBIjANBgkqhkiG9w0BAQEFAAOCAQ8AMIIBCgKCAQEArsM5EJyfZD4tNO+e\
+CTX/qMXGWUpsfPtub8mldH27QTt+nuLWXMAywtZ0vWfT4w3yxif53bQNNv6H+koCujQGn7o9CqUl\
+QOY3GgDX3MGxLKNhvwKxCPm+XChtduHi69khyKVSLwCtIFCVvKaQNJsyI5I7i06V1RqyQW7tOSfw\
+oCo+0sZGRz24OdD/0skPguxbPl9KYGvKfUDwRboHBf82bwaRisT7ASexLWecWxA/XKlq9v/irc2n\
+Op4ofBfb5d2KmdGUk49/NgHr66TvD4WPec075HocJ3aYizg9EtRrZD52hqY++HszhbJiOY1fm0Xo\
+JrUn5TeHtlVHGygX02ggPQIDAQABoy8wLTAMBgNVHRMEBTADAQH/MB0GA1UdDgQWBBSAoKfrUYK8\
+B/yaJSQ+s93WrhWVizANBgkqhkiG9w0BAQsFAAOCAQEARoizzGqOEluCHXMhXg039D0mK6JQHuCD\
+8i71GJWV7imhgb/aH4Q2UgHOzgBi8eVi4o9oIuHPIDIu6Un1DSPvVxBBCeYxjPSRXBYzD5HEl/F6\
+t3mqOg38AXu0/CO5ZcZ9dhc+nPF04wwqkPUa5GHqdx6vXc/nnodGX+0q2IToaioYRECvwJwluxWP\
+qL48Ta8jXIKDCIVNpHoHfe6EhSx6WTp5+LQaiNrUHrTwvEhA62CkMNlXrXBIwrZd5AQNtnUGEQsG\
+wM4fcy9YAD2F3yQJxv5dDVAk1dmtr6PapPYv+wqtoHR3AO2i7IuegX1XLvMm+5XES8ECzpz6pWaf\
+/wLEnCJEEiEIgNTb0g8SGXapFO2NfiXoI85ijPezGjexnyq3JUqKiKwYvNnmnAYgvN27/REqADIR\
+dGVzdG1lcmNoYW50OC5vcmcqgAJ85QFFRausmWQmjhLiG+mES/Nvq1kHlwzYZaEi6sFbFPQgN1xP\
+NRn4BOUUws+izkB/fsyWYOHhOPee507G4SEoW9DyzCpcovRoY+iFl1Ro6EYnjHoBsnzuhbHKaNxy\
+v9gJ40pgCsJjMlrk48QA1yK82NcaYYmR/+mhlxzQt1Z9i/i2mVi1DLWooqlZL2brxKVFY0OvgUZn\
+xA6wlRViEgCsonG9s05HhyWlOfULGubBo17zeY57ujG+R20Hb3vXw+P+po5512RYT0SGIaYCmVvI\
+mdCEQpUOI7/0Omw9i3piuw7jTZ8i6kmp6gIZe3lBbzKJ3IzvfDL6A1O6SMJg/5Lu\
 ";
 
 //
@@ -431,7 +496,7 @@ d2hj739GDLz0b5KuJ2SG6VknMRQM976w/m2qlq0ccVGaaZ2zMIGfpzL3p6adwx/5\
 
 //
 // Payment request with amount overflow (amount is set to
-// 21,000,001 XPI)
+// 210,000,010,000 XPI)
 //
 const char *paymentrequest5_cert2_BASE64 = "\
 Egt4NTA5K3NoYTI1NhrQBArNBDCCAkkwggExoAMCAQICAQEwDQYJKoZIhvcNAQEL\
diff --git a/test/functional/logos_feature_minerfund.py b/test/functional/logos_feature_minerfund.py
index 114af63b9..68da6991a 100755
--- a/test/functional/logos_feature_minerfund.py
+++ b/test/functional/logos_feature_minerfund.py
@@ -72,24 +72,14 @@ class MinerFundTest(BitcoinTestFramework):
         prepare_block(block)
         assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
 
-        # Submit a custom block that does send the required outputs
-        # with the miner reward output in the middle (which is allowed)
-        # but with the outputs in the wrong order (which is not allowed)
+        # Submit a custom block that sends the required outputs
+        # with the miner reward at the end (which is allowed)
         coinbase.vout[1:] = []
         for output in expected_outputs:
             coinbase.vout.append(CTxOut(int(output['value']),
                                         CScript(bytes.fromhex(output['scriptPubKey']))))
         coinbase.vout.insert(2, CTxOut(int(SUBSIDY / 2 * COIN),
                                        CScript([OP_HASH160, bytes(20), OP_EQUAL])))
-        coinbase.vout[2], coinbase.vout[4] = coinbase.vout[4], coinbase.vout[2]
-        coinbase.rehash()
-        block = create_block(int(best_block_hash, 16),
-                             coinbase, block_height, block_time + 1)
-        prepare_block(block)
-        assert_equal(node.submitblock(ToHex(block)), 'bad-cb-minerfund')
-
-        # Fix the order and block is valid
-        coinbase.vout[2], coinbase.vout[4] = coinbase.vout[4], coinbase.vout[2]
         coinbase.rehash()
         block = create_block(int(best_block_hash, 16),
                              coinbase, block_height, block_time + 1)
@@ -113,6 +103,8 @@ class MinerFundTest(BitcoinTestFramework):
         coinbase = create_coinbase(block_height)
         coinbase.vout[1] = CTxOut(int(share_multiplier * SUBSIDY / 2 * COIN),
                                   CScript([OP_HASH160, bytes(20), OP_EQUAL]))
+        block_template = node.getblocktemplate()
+        expected_outputs = block_template['coinbasetxn']['minerfund']['outputs']
         for output in expected_outputs:
             coinbase.vout.append(CTxOut(int(share_multiplier * output['value']),
                                         CScript(bytes.fromhex(output['scriptPubKey']))))
-- 
2.39.5 (Apple Git-154)

